src/Setup/Data/VersionFiles.purs

```diff
@@ -0,0 +1,127 @@
+  | VersionParseError String ParseError
+  | ToolNameError JsonDecodeError
+
+printV2FileError :: V2FileError -> String
+printV2FileError = case _ of
+  JsonCodecError e -> printJsonDecodeError e
+  VersionParseError field e -> fold
+    [ "Version parse failure for key, "
+    , field
+    , "': "
+    , show e
+    ]
+  ToolNameError e -> fold
+    [ "Unable to convert String into Tool. "
+    , printJsonDecodeError e
+    ]
+
+type LatestUnstable a =
+  { latest :: a
+  , unstable :: a
+  }
+
+newtype V2FileSchema = V2FileSchema
+  { fileUrl :: String
+  , localFile :: String
+  , encode :: Map Tool (LatestUnstable Version) -> Json
+  , decode :: Json -> Either V2FileError (Map Tool (LatestUnstable Version))
+  }
+
+derive instance Newtype V2FileSchema _
+
+version2 :: V2FileSchema
+version2 = V2FileSchema
+  { fileUrl: "https://raw.githubusercontent.com/purescript-contrib/setup-purescript/main/dist/versions" <> vSuffix <> ".json"
+  , localFile: "./dist/version" <> vSuffix <> ".json"
+  , encode: foldlWithIndex encodeFoldFn jsonEmptyObject
+  , decode: \j -> do
+      obj :: Object Json <- lmap JsonCodecError $ decodeJson j
+      keyVals <- forWithIndex obj \key val -> do
+        tool <- strToTool key
+        { latest
+        , unstable
+        } :: LatestUnstable String <- lmap JsonCodecError $ decodeJson val
+        latest' <- lmap (VersionParseError (key <> ".latest")) $ Version.parseVersion latest
+        unstable' <- lmap (VersionParseError (key <> ".unstable")) $ Version.parseVersion unstable
+        pure $ Tuple tool { latest: latest', unstable: unstable' }
+      pure $ Map.fromFoldable keyVals
+  }
+  where
+  vSuffix = "-v2"
+  encodeFoldFn tool acc { latest, unstable }
+    | Just toolStr <- toolToMbString tool = do
+        let rec = { latest: Version.showVersion latest, unstable: Version.showVersion unstable }
+        toolStr := rec ~> acc
+    | otherwise =
+        acc
+
+  -- in case we add support for other tools in the future...
+  toolToMbString = case _ of
+    PureScript -> Just "purs"
+    Spago -> Just "spago"
+    Psa -> Just "psa"
+    PursTidy -> Just "purs-tidy"
+    Zephyr -> Just "zephyr"
+
+  strToTool = case _ of
+    "purs" -> Right PureScript
+    "spago" -> Right Spago
+    "psa" -> Right Psa
+    "purs-tidy" -> Right PursTidy
+    "zephyr" -> Right Zephyr
+    str -> Left $ ToolNameError $ UnexpectedValue $ encodeJson str
+
+newtype V1FileSchema = V1FileSchema
+  { localFile :: String
+  , encode :: Map Tool Version -> Json
+  }
+
+derive instance Newtype V1FileSchema _
+
+version1 :: V1FileSchema
+version1 = V1FileSchema
+  { localFile: "./dist/versions.json"
+  , encode: foldlWithIndex encodeFoldFn jsonEmptyObject
+  }
+  where
+  encodeFoldFn tool acc version
+    | Just toolStr <- toolToMbString tool =
+        toolStr := Version.showVersion version ~> acc
+    | otherwise =
+        acc
+
+  -- in case we add support for other tools in the future...
+  toolToMbString = case _ of
```

Comment:
Maybe just `printTool` for the name, and update the comment to say something like:

```
-- We preserve the set of tools that existed at the time this version format was produced;
-- if more tools are added, they should map to `Nothing`
```